static/app/utils/useUndoableReducer.tsx

```diff
@@ -0,0 +1,88 @@
+import {ReducerAction, ReducerState, useReducer} from 'react';
+
+export type UndoableNode<S> = {
+  current: S;
+  next: UndoableNode<S> | undefined;
+  previous: UndoableNode<S> | undefined;
+};
+
+type UndoAction = {
+  type: 'undo';
+};
+
+type RedoAction = {
+  type: 'redo';
+};
+
+type UndoableReducerAction<A> = UndoAction | RedoAction | A;
+
+function isUndoOrRedoAction(
+  action: UndoableReducerAction<any>
+): action is UndoAction | RedoAction {
+  if (action.type) {
+    return action.type === 'undo' || action.type === 'redo';
+  }
+  return false;
+}
+
+function undoableReducer<S, A>(
+  state: UndoableNode<S>,
+  action: UndoableReducerAction<A>
+): UndoableNode<S> {
+  if (!isUndoOrRedoAction(action)) {
+    return state;
+  }
+
+  if (action.type === 'undo') {
+    return state.previous ? state.previous : state;
+  }
+
+  if (action.type === 'redo') {
+    return state.next ? state.next : state;
+  }
+
+  throw new Error('Unreachable case');
+}
+
+type CombinedReducer<R extends React.Reducer<any, any>> = React.Reducer<
+  UndoableNode<ReducerState<R>>,
+  UndoableReducerAction<ReducerAction<R>>
+>;
+
+export function makeUndoableReducer<R extends React.Reducer<any, any>>(
+  reducer: R
+): CombinedReducer<R> {
+  return (
+    state: UndoableNode<ReducerState<R>>,
+    action: UndoableReducerAction<ReducerAction<R>>
+  ) => {
+    const maybeUndoOrRedo = undoableReducer(state, action);
+
+    if (maybeUndoOrRedo !== state) {
+      return maybeUndoOrRedo;
```

Comment:
Isn't this just returning the new state when the reducer is expecting an object with like

```typescript
{
  next: ...,
  current: ...,
  previous: ...,
}
```

Reply:
Yeah, this wasnt correct :D I added proper tests later and fixed it all